design/namespaced-roles.md

```diff
@@ -0,0 +1,299 @@
+
+In order to have this flexibility and still support current use cases, we could create a new kind of role or backport existing roles to this new format: Namespaced roles.
+
+## Specification
+
+A Namespaced Role would follow a certain structure that would unambiguously identify it across the system, following a hierarchical structure.
+
+They would work similarly to ARNs in AWS, where a portion of the format is fixed and always identifiable by position, with the possibility of containing empty fields, containing a final variable part that could be freely named.
+
+An example would be:
+
+`role_v1:/\<realm\>/\<tenant\>/\<client\>/\<group\>/<roleName>`
+
+**Realm**: All resources in Keycloak are namespaced to a realm, so a unique identifier for a role would also contain it.
+   - It doesn't support empty values - all roles will be scoped to a realm.  
+   - Example: `role_v1:/realm1////roleName`
+
+**Tenant**: There's a proposal to make Keycloak multitenant aware, so even if it's not currently supported, for future compatibility, it should be considered when defining a unique identifier.
+   - Supports empty values - A role might be created and assigned at a realm level
+   - Example: `role_v1:/realm1/tenantA///roleName`, `role_v1:/realm1///groupA/roleName`
+
+**Client**: Roles can be defined and assigned for a certain client in Keycloak, so a unique identifier has to support this possibility.
+   - Supports empty values - A role might be created and assigned without being restricted to a client.
+   - Example: `role_v1:/realm1/tenantA/clientId//roleName`, `role_v1:/realm1/tenantA//groupA/roleName`
+
+**Group**: Roles can be assigned to groups or, as intended in this design proposal, assigned to a user in the context of a group.
+   - Supports empty values - A role might be created at any level and not necessarily assigned to a group.
+   - Example: `role_v1:/realm1/tenantA/clientId/groupB/roleName`, `role_v1:/realm1/tenantA/clientId//roleName`
+
+**RoleName**: Finally, the name of the role that is being defined.
+   - Doesn't support empty values - You need to know the name of the role that is being created.
+   - Example: `role_v1:/realm1/tenantA/clientId/groupB/roleName`
+
+A possible free-form role identifier could be allowed for resource-server specific roles.
+
+The initial fixed part would be kept in order to identify the role within a realm, tenant, client, group etc, but instead of a fixed name for the role, it would be a reference to other hierarchical role formed with any random structure.
+
+A possible format would be:
+
+- `role_v1:/realm1/tenant1//group1/frole_v1:/mycompany/resources/department-a-roles/developer`
+
+If the role name is another reference to a free form role, starting with `frole_v1:/`, this role will also be represented in a hierarchical way, making it possible for protocol mappers to show them as such in tokens.
+
+## Use Cases
```

Comment:
Would be good to explicitly mention how role namespaces can clean-up the roles used "internally" by Keycloak's admin APIs. Getting rid of the strange clients, and having a separate/dedicated namespace for the roles so they don't conflict with "custom roles". I had something like the following in mind:
```
kc/admin/{realm-name}/{role-name}
```